### DISCUSSION TITLE
How to subscribe to the properties of tree child nodes?

### DISCUSSION THREAD

**Discussion by NicksGhost** [reactiveui/ReactiveUI#3715] [Q&A] [answered]
Excuse me, I want to know how to subscribe to the attribute changes of any child node of the tree. I don't seem to find it in the documentation.

```
public class NodeViewModel : ReactiveObject
{
    [Reactive]
    public bool IsChecked { get; set; }

    [Reactive]
    public string Name { get; set; }

    public ObservableCollection<NodeViewModel>? Children { get; set; }
}
```

Just like the above code structure, I hope that no matter how deep the recursive structure is, I can observe the changes in IsChecked. How to do this?

```
public class TreeViewModel : ReactiveObject
{
    public readonly ObservableCollection<NodeViewModel> Nodes = [];

    public TreeViewModel () {
        // FIXME: In this way, I can only get the first level of IsChecked. How to operate deeper?
        this.WhenAnyValue(o => o.Nodes.IsChecked)
            .Subscribe(o => {
                  Console.WriteLine(o);
            });
    }
}
```

**Comment by anaisbetts** [comment, marked as answer]
There's no super easy way to do this at the moment. The short version for a static list looks something like (coding via TextArea, forgive typos please!):

```cs
var anyCheckedChanged = Observable.Merge(
  Nodes.Select(node => 
    node.WhenAnyValue(x => x.IsChecked).Skip(1).Select(checked => (node, checked))));
```

(we skip 1 because we don't want to fire N initial states, though maybe you might want to do that!)

The problem is that you need to rebuild this Observable whenever any items in the list change. This is where DynamicData is a good solution, to create a projection of your IsChecked values. Another solution is building on our previous solution, something like:

```cs
var anyCheckedChange = listItemsAddedRemovedOrReplaced.StartWith(null)
  .Select(_ => {
    var anyCheckedChangedForThisParticularList = Observable.Merge(
      Nodes.ToArray().Select(node => 
        node.WhenAnyValue(x => x.IsChecked).Skip(1).Select(checked => (node, checked))));
    
    return anyCheckedChangedForThisParticularList;
  })
  .Switch();
```  

If the list is constantly changing multiple times a second, this is a pretty inefficient solution, but for most everything else it's Fine

**Reply by NicksGhost** [reply]
Can this problem be solved using MessageBus?

**Reply by anaisbetts** [reply]
I'm not sure how MessageBus would help here

**Comment by ChrisPulman** [comment]
I have a Treeview example that may assist https://github.com/ChrisPulman/ReactiveUITreeviewExample

**Reply by ChrisPulman** [reply]
The difference between your snippet and mine is, mine is using the IsSelected Property and yours IsChecked.
You should be able to modify my code to cater for your needs, by using FlattenAndSelect you can select the relevant properties as desired.
The ReactiveList is based on DynamicData and has CurrentItems, AddedItems, and RemovedItems Observables, the TreeView example only uses CurrentItems and Items, but you could modify to suit your needs if you need Added or Removed items.

**Reply by NicksGhost** [reply]
Thanks for your example, very useful to me.
